fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/translation/MaxMatchMap.java

```diff
@@ -102,117 +224,789 @@ public ValueEquivalence getValueEquivalence() {
+                                                                   @Nonnull final Deque<IncrementalValueMatcher> matchers,
+                                                                   final int maxDepthBound,
+                                                                   @Nonnull final Set<Value> expandedValues) {
+        if (maxDepthBound == 0) {
+            return ImmutableMap.of(currentQueryValue, MatchResult.notMatched());
+        }
+
+        //
+        // Localize all given matchers and add one for this level. Keep a boolean variable indicating if there are
+        // any potential parent matchers that can still match something in the subtree.
+        //
+        final var localMatchers = new ArrayDeque<IncrementalValueMatcher>();
+        boolean anyParentsMatching = false;
+        if (descendOrdinal >= 0) {
+            for (final var matcher : matchers) {
+                anyParentsMatching |= matcher.anyMatches();
+                final var descendedMatcher = matcher.descend(currentQueryValue, descendOrdinal);
+                localMatchers.addLast(descendedMatcher);
+            }
+        }
+
+        if (!anyParentsMatching) {
+            if (knownValueMap.containsKey(currentQueryValue)) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("getting memoized info value={}", currentQueryValue);
+                }
+                return knownValueMap.get(currentQueryValue);
+            }
+        }
+
+        //
+        // If any parents have any potential matches, we need to generate all possible expansions for this subtree as
+        // we need to exhaustively find the case where the entire subtree rooted in one of the parents matches. If no
+        // such potential matches exist, we know that a match in this frame immediately also is the maximum match.
+        //
+
+        //
+        // Create a matcher for this level.
+        //
+        final var currentMatcher =
+                IncrementalValueMatcher.initial(currentQueryValue, candidateValue);
+        localMatchers.push(currentMatcher);
+        final var isCurrentMatching = currentMatcher.anyMatches();
+
+        //
+        // Keep a results map that maps from a Value variant to a match result. Note that if we have don't have
+        // a potential parent match we can make local decisions in this method, specifically, we can prune the variants
+        // that are being generated early on.
+        //
+        final var bestMatches = new BestMatches(currentQueryValue, !anyParentsMatching);
+
+        final var children = currentQueryValue.getChildren();
+        if (Iterables.isEmpty(children)) {
+            final var resultForCurrent =
+                    computeForCurrent(maxDepthBound, currentQueryValue, candidateValue, rangedOverAliases,
+                            valueEquivalence, ImmutableList.of());
+            bestMatches.put(currentQueryValue, resultForCurrent);
```

Comment:
I think this is where we're performing the actual binding, i.e. if we find a new potential match we'll add it, but only if it within bounds, i.e. its depth is less than `currentMaxDepth`, right?

Reply:
Yeah, here and a little further down when we have a query value that has children.